relayer/chain/substrate/outgoing_extrinsics.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2020 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package substrate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/snowfork/go-substrate-rpc-client/v2/types"
+	"golang.org/x/sync/errgroup"
+)
+
+const MaxWatchedExtrinsics = 500
+
+type extrinsicPool struct {
+	sync.Mutex
+	conn     *Connection
+	eg       *errgroup.Group
+	log      *logrus.Entry
+	maxNonce uint32
+	watched  uint
+}
+
+func newExtrinsicPool(eg *errgroup.Group, conn *Connection, log *logrus.Entry) *extrinsicPool {
+	ep := extrinsicPool{
+		conn:    conn,
+		eg:      eg,
+		log:     log,
+		watched: 0,
+	}
+	return &ep
+}
+
+func (ep *extrinsicPool) WaitForSubmitAndWatch(ctx context.Context, nonce uint32, ext *types.Extrinsic) {
+	defer ep.Unlock()
+
+	for {
+		ep.Lock()
+		if ep.hasCapacity() {
+			ep.eg.Go(func() error {
+				return ep.submitAndWatchLoop(ctx, nonce, ext)
+			})
+
+			ep.watched++
+			return
+		}
+		ep.Unlock()
+
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+func (ep *extrinsicPool) hasCapacity() bool {
+	return ep.watched < MaxWatchedExtrinsics
+}
+
+func (ep *extrinsicPool) submitAndWatchLoop(ctx context.Context, nonce uint32, ext *types.Extrinsic) error {
+	sub, err := ep.conn.api.RPC.Author.SubmitAndWatchExtrinsic(*ext)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("Context was canceled. Stopping extrinsic monitoring")
+
+		case status := <-sub.Chan():
+			if status.IsDropped || status.IsInvalid {
+				// Indicates that the extrinsic wasn't processed. We expect the Substrate txpool to be
+				// stuck until this nonce is successfully provided. But it might be provided without this
+				// relayer's intervention, e.g. if an internal Substrate mechanism re-introduces it to the
+				// txpool.
+				sub.Unsubscribe()
+				statusStr := getStatusString(&status)
+				ep.log.WithFields(logrus.Fields{
+					"nonce":  nonce,
+					"status": statusStr,
+				}).Debug("Extrinsic failed to be processed")
+
+				// Back off to give the txpool time to resolve any backlog
+				time.Sleep(ep.getRetryDelay(nonce))
+
+				ep.Lock()
+				if nonce <= ep.maxNonce {
+					// We're in the clear - no need to retry
+					ep.watched--
+					ep.Unlock()
+					return nil
+				}
+				ep.Unlock()
+
+				// This might fail because the transaction has been temporarily banned in Substrate. In that
+				// case it's best to crash, wait a while and restart the relayer.
+				ep.log.WithFields(logrus.Fields{
+					"nonce":  nonce,
+					"status": statusStr,
+				}).Debug("Re-submitting failed extrinsic")
+				newSub, err := ep.conn.api.RPC.Author.SubmitAndWatchExtrinsic(*ext)
+				if err != nil {
+					return err
+				}
+				sub = newSub
+
+			} else if !status.IsReady && !status.IsFuture && !status.IsBroadcast {
```

Comment:
Would it make sense to check for `Finalized` or `FinalityTimeout` instead?

Reply:
Maybe? My aim was for `extrinsicPool` to guarantee that Substrate processes all nonces without gaps. So `InBlock` is good enough. But it might make sense for the pool to not only guarantee that the nonce is provided, but additionally guarantee that the extrinsic is included in a block. Which do you think makes more sense?

Reply:
Sorry, have missed your ping here. I believe the relayer should fail load and noisy if an extrinsic fails, but for the pool it should be fine to just guarantee nonce processing 👍